File: 4.8-bc/ts/stage-2-feature-first/src/features/catalog/handler.ts
import type { FastifyInstance } from "fastify";

import { writeError } from "../../shared/httputil.js";
import type { Service } from "./service.js";

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.get("/products", async (req, reply) => {
      try {
        const products = await this.svc.list();
        return reply.code(200).send({
          products: products.map((p) => ({
            id: p.id,
            name: p.name,
            price_cents: p.priceCents,
            category_id: p.categoryId,
          })),
        });
      } catch (err) {
        req.log.error({ err }, "list products failed");
        return writeError(reply, err);
      }
    });
  }
}

File: 4.8-bc/py/stage-2-feature-first/app/features/catalog/handler.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.catalog.repository import Repository
from app.features.catalog.service import Service
from app.shared.db import get_session

router = APIRouter(tags=["catalog"])


class ProductDTO(BaseModel):
    id: str
    name: str
    price_cents: int
    category_id: str


class ProductsResp(BaseModel):
    products: list[ProductDTO]


@router.get("/products", response_model=ProductsResp)
async def list_products(
    session: AsyncSession = Depends(get_session),
) -> ProductsResp:
    svc = Service(Repository(session))
    products = await svc.list()
    return ProductsResp(
        products=[
            ProductDTO(
                id=str(p.id),
                name=p.name,
                price_cents=p.price_cents,
                category_id=str(p.category_id),
            )
            for p in products
        ]
    )
